frontend/apps/filemanager/filemanager.lua

```diff
@@ -88,7 +88,7 @@ function FileManager:init()
         padding_left = Size.padding.large,
         padding_right = Size.padding.large,
         padding_bottom = 0,
-        callback = function() self:tapPlus() end,
+        callback = nil, -- top right corner callback handled by gesture manager
```

Comment:
I could take it back out again. Maybe I'll think of something after a night's sleep. ;-)

Reply:
Btw, not really a first. This is basically exactly what the flipping & bookmark corners do.

Reply:
They do not feel like that to me. ReaderDogear and ReaderFlipping widgets both are painted (and so are containers of their own inner widgets), set the tap zone, and handle the action.

Reply:
They might technically set the tap zone, but the tap zone they set is completely unrelated to the actual widget.

https://github.com/koreader/koreader/blob/22db71c07549ffc59196f1f27cd9bb162eacd6d7/frontend/apps/reader/modules/readerflipping.lua#L32-L46

https://github.com/koreader/koreader/blob/22db71c07549ffc59196f1f27cd9bb162eacd6d7/frontend/apps/reader/modules/readerdogear.lua#L77-L91

Reply:
I was not referring about tap zone overlaping with the icon. Just the self containment of the 3 parts involved (widget, tap zone, action) in a single InputContainer.
What bothers me is the "remote overlay", the tap zone being set far away in readergesture.
But now, it's no more a first, but a second and a third :)
But again, I don't see how else.

Reply:
True, it's nice that they're self-contained units.

I think the "proper" way would be to allow the user to set their own icon+action+possibly tap zone in place (including one that's transparent) so that they are more together.

In Opera's toolbar.ini that'd look like this:
```
[Document Toolbar.content]
NamedButton0, tbb_Back, -847360260="Back + Show hidden popup menu, "Back Menu""
NamedButton1, tbb_Forward, 830055518="Forward + Show hidden popup menu, "Internal Forward History""
Button2, -108388079="Fast Forward, , , "Next" + Show hidden popup menu, "Internal Fast Forward History""
Button3, -119414254="Wand | Fast Forward, , , "Next" + Show hidden popup menu, "Internal Fast Forward History""
Address4
Button5, "''InO"="Execute program, "opera-developer", "%u", "InO", "OperaBlink""
Button6, "InFx"="Execute program, "firefox", "%u", "InFx", "Firefox""
Zoom7
ExtensionSet8
```

In Otter it's done with JSON:

```js
[
	{
		"actions": [
			"RewindAction",
			"GoBackAction",
			"GoForwardAction",
			"FastForwardAction",
			"ReloadOrStopAction",
			"AddressWidget",
			"SearchWidget"
		],
		"buttonStyle": "iconOnly",
		"fullScreenVisibility": "hidden",
		"identifier": "AddressBar",
		"normalVisibility": "visible",
		"row": 0,
		"title": "Address Bar"
	},
	{
		"buttonStyle": "iconOnly",
		"currentPanel": "bookmarks",
		"fullScreenVisibility": "hidden",
		"identifier": "SideBar",
		"location": "left",
		"normalVisibility": "visible",
		"panelSize": 530,
		"panels": [
			"bookmarks",
			"cookies",
			"history",
			"notes",
			"passwords",
			"transfers",
			"windows",
			"config"
		],
		"row": 0,
		"title": "Sidebar"
	}
]
```

That way you've got clear separation of concern in the code but everything is also clearly defined together and configurable. We've currently got a bit of an awkward in-between.

Also I don't have a clear vision yet for the outcome, but I do know that with all the actions that've been added to the gesture manager it's much easier to move to a different/further abstraction than it was a few months ago.

Reply:
Well, I hope we don't need all that :)
Also, we're in a bit different situation as the code is in Lua, so we don't really need to provide as much hooks for customisation as the apps in compiled C that are not hackable - the interested users (very little) can always hack/patch/monkey-patch our code base to have things a bit more to their liking (I do that). Complexifying and abstracting with different levels of indirection to satisfy a small user base is not really to my liking :)

But yes, with gesturemanager, it has become different - you could possibly move everything reader/filemanager gestures in there to have them in a single place. But that would be one global level of indirection that would make reading/hacking the code a bit more tedious.

Reply:
Architecturally speaking it may be better for the gesture manager to be more like the plugin manager. Actually it's already quite a bit like that with the event system. But anyway, it's working and I have no special desire to waste time redesigning it or whatnot. ;-)